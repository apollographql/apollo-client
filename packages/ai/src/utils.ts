import {
  GraphQLInputObjectType,
  GraphQLInputType,
  GraphQLObjectType,
  GraphQLOutputType,
  GraphQLUnionType,
  InputObjectTypeDefinitionNode,
  isListType,
  isWrappingType,
  Kind,
  ObjectTypeDefinitionNode,
  TypeNode,
  UnionTypeDefinitionNode,
} from "graphql";

export type NamedNode = { name: { value: string } };

/**
 * The names of the root types in the GraphQLSchema.
 */
export enum RootTypeName {
  MUTATION = "Mutation",
  QUERY = "Query",
  SUBSCRIPTION = "Subscription",
}

/**
 * The sort order of the root types in the schema.
 */
const ROOT_TYPE_ORDER: { [key: string]: number } = {
  [RootTypeName.QUERY]: 0,
  [RootTypeName.MUTATION]: 1,
  [RootTypeName.SUBSCRIPTION]: 2,
};

/**
 * Check if a number is a float (i.e. 9.5).
 *
 * @param num - The number to check.
 * @returns True if the number is a float, false otherwise.
 */
export function isFloat(num: number) {
  return typeof num === "number" && !Number.isInteger(num);
}

/**
 * Convert a plural word to its singular form.
 *
 * @param str - The plural word to convert.
 * @returns The singular form of the word.
 */
export function singularize(str: string) {
  if (!str) {
    return "";
  }

  // Handle common pluralization patterns
  if (str.endsWith("ies")) {
    return str.slice(0, -3) + "y";
  } else if (str.endsWith("ves")) {
    return str.slice(0, -3) + "f";
  } else if (str.endsWith("es")) {
    // Special cases for -es endings
    if (str.endsWith("ches") || str.endsWith("shes") || str.endsWith("xes")) {
      return str.slice(0, -2);
    } else if (str.endsWith("ses")) {
      return str.slice(0, -2);
    } else {
      return str.slice(0, -1);
    }
  } else if (str.endsWith("s") && str.length > 1) {
    return str.slice(0, -1);
  }

  return str;
}

/**
 * Convert the first letter of a string to uppercase.
 *
 * @param str - The string to convert.
 * @returns The string with the first letter capitalized.
 */
export function ucFirst(str: string) {
  if (!str) {
    return "";
  }
  return str.charAt(0).toUpperCase() + str.slice(1);
}

/**
 * Sorts object-level AST nodes by their name.
 *
 * AST Nodes named after a root type (like Query, Mutation, Subscription) are
 * sorted to the beginning of the list based on their defined order.
 *
 * All other AST Nodes are sorted alphabetically regardless of kind.
 * @param list — The list of AST Nodes to sort.
 * @returns The sorted list of AST Nodes.
 */
export function sortObjectASTNodes<T extends NamedNode>(list: T[]): T[] {
  return list.sort((a, b) => {
    const aName = a.name.value;
    const bName = b.name.value;

    const aOrder = ROOT_TYPE_ORDER[aName];
    const bOrder = ROOT_TYPE_ORDER[bName];

    // If both are root types, sort by their defined order
    if (aOrder !== undefined && bOrder !== undefined) {
      return aOrder - bOrder;
    }

    // If only a is a root type, it goes first
    if (aOrder !== undefined) {
      return -1;
    }

    // If only b is a root type, it goes first
    if (bOrder !== undefined) {
      return 1;
    }

    // Neither is a root type, sort alphabetically
    return aName.localeCompare(bName);
  });
}

/**
 * Sorts union member names alphabetically.
 * @param members — The list of union member names to sort.
 * @returns The sorted list of union member names.
 */
export function sortUnionMembers(members: string[]): string[] {
  return members.sort((a, b) => {
    return a.localeCompare(b);
  });
}

/**
 * Sorts AST nodes by their name alphabetically.
 *
 * Unlike `sortObjectASTNodes`, this function sorts alphabetically regardless
 * of kind or name.
 * @param nodes — The list of AST nodes to sort.
 * @returns The sorted list of AST nodes.
 */
export function sortASTNodes<T extends NamedNode>(nodes: T[]): T[] {
  return nodes.sort((a, b) => {
    return a.name.value.localeCompare(b.name.value);
  });
}

/**
 * Transforms a GraphQL Input/Output type to an AST TypeNode.
 * @param type — The GraphQL Input/Output type to transform
 * @returns The AST TypeNode
 */
export function graphQLTypeToTypeNode(
  type: GraphQLOutputType | GraphQLInputType
): TypeNode {
  if (isWrappingType(type)) {
    // Recursively transform the wrapped type
    const returnType = graphQLTypeToTypeNode(type.ofType);

    // If the type is a non-null type, return the non-null type to avoid
    // unnecessary wrapping (but this also shouldn't actually happen)
    if (returnType.kind === Kind.NON_NULL_TYPE) {
      return returnType;
    }

    // Return the wrapped type
    return {
      kind: isListType(type) ? Kind.LIST_TYPE : Kind.NON_NULL_TYPE,
      type: returnType,
    };
  }
  return {
    kind: Kind.NAMED_TYPE,
    name: { kind: Kind.NAME, value: type.name },
  };
}

/**
 * Transforms a GraphQL Object type to an AST ObjectTypeDefinitionNode.
 * @param type — The GraphQL Object type to transform
 * @returns The AST ObjectTypeDefinitionNode
 */
export function graphQLObjectTypeToObjectTypeDefinitionNode(
  type: GraphQLObjectType
): ObjectTypeDefinitionNode {
  return {
    kind: Kind.OBJECT_TYPE_DEFINITION,
    name: { kind: Kind.NAME, value: type.name },
    fields: sortASTNodes(
      Object.values(type.getFields()).map((field) => ({
        kind: Kind.FIELD_DEFINITION,
        name: { kind: Kind.NAME, value: field.name },
        type: graphQLTypeToTypeNode(field.type),
      }))
    ),
  };
}

/**
 * Transforms a GraphQL Input type to an AST InputObjectTypeDefinitionNode.
 * @param type — The GraphQL Input type to transform
 * @returns The AST InputObjectTypeDefinitionNode
 */
export function graphQLInputObjectTypeToInputObjectDefinitionNode(
  type: GraphQLInputObjectType
): InputObjectTypeDefinitionNode {
  return {
    kind: Kind.INPUT_OBJECT_TYPE_DEFINITION,
    name: { kind: Kind.NAME, value: type.name },
    fields: sortASTNodes(
      Object.values(type.getFields()).map((field) => ({
        kind: Kind.INPUT_VALUE_DEFINITION,
        name: { kind: Kind.NAME, value: field.name },
        type: graphQLTypeToTypeNode(field.type),
      }))
    ),
  };
}

/**
 * Transforms a GraphQL Union type to an AST UnionTypeDefinitionNode.
 * @param type — The GraphQL Union type to transform
 * @returns The AST UnionTypeDefinitionNode
 */
export function graphQLUnionTypeToUnionTypeDefinitionNode(
  type: GraphQLUnionType
): UnionTypeDefinitionNode {
  return {
    kind: Kind.UNION_TYPE_DEFINITION,
    name: { kind: Kind.NAME, value: type.name },
    types: sortASTNodes(
      Object.values(type.getTypes()).map((memberType) => ({
        kind: Kind.NAMED_TYPE,
        name: { kind: Kind.NAME, value: memberType.name },
      }))
    ),
  };
}

/**
 * Deep merge utility function to preserve nested properties.
 *
 * @param target - The target object to merge into.
 * @param source - The source object to merge from.
 * @returns The merged object.
 */
export function deepMerge(target: any, source: any): any {
  if (source === null || typeof source !== "object") {
    return source;
  }

  if (Array.isArray(source)) {
    return source;
  }

  if (target === null || typeof target !== "object" || Array.isArray(target)) {
    target = {};
  }

  const result = { ...target };

  for (const key in source) {
    if (source.hasOwnProperty(key)) {
      if (
        typeof source[key] === "object" &&
        source[key] !== null &&
        !Array.isArray(source[key])
      ) {
        result[key] = deepMerge(result[key], source[key]);
      } else {
        result[key] = source[key];
      }
    }
  }

  return result;
}

/**
 * Converts a value to a formattedJSON string.
 * This handles Map and Set instances by converting them to objects/arrays.
 * @param value — The value to convert to a JSON string
 * @returns The JSON string
 */
export function toJSON(value: any) {
  return JSON.stringify(
    value,
    (key: string, value: any) => {
      if (value instanceof Map) {
        return {
          dataType: "Map",
          value: Object.fromEntries(value.entries()),
        };
      } else if (value instanceof Set) {
        return {
          dataType: "Set",
          value: [...value],
        };
      } else {
        return value;
      }
    },
    2
  );
}
