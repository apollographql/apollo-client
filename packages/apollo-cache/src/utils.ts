import {
  DocumentNode,
  OperationDefinitionNode,
  SelectionSetNode,
  FieldNode,
  FragmentDefinitionNode,
} from 'graphql';

export function queryFromPojo(obj: any): DocumentNode {
  const op: OperationDefinitionNode = {
    kind: 'OperationDefinition',
    operation: 'query',
    name: {
      kind: 'Name',
      value: 'GeneratedClientQuery',
    },
    selectionSet: selectionSetFromObj(obj),
  };

  const out: DocumentNode = {
    kind: 'Document',
    definitions: [op],
  };

  return out;
}

export function fragmentFromPojo(obj: any, typename?: string): DocumentNode {
  const frag: FragmentDefinitionNode = {
    kind: 'FragmentDefinition',
    typeCondition: {
      kind: 'NamedType',
      name: {
        kind: 'Name',
        value: typename || '__FakeType',
      },
    },
    name: {
      kind: 'Name',
      value: 'GeneratedClientQuery',
    },
    selectionSet: selectionSetFromObj(obj),
  };

  const out: DocumentNode = {
    kind: 'Document',
    definitions: [frag],
  };

  return out;
}

function selectionSetFromObj(obj: any): SelectionSetNode {
  if (
    typeof obj === 'number' ||
    typeof obj === 'boolean' ||
    typeof obj === 'string' ||
    typeof obj === 'undefined' ||
    obj === null
  ) {
    // No selection set here
    return null;
  }

  if (Array.isArray(obj)) {
    // GraphQL queries don't include arrays
    return selectionSetFromObj(obj[0]);
  }

  // Now we know it's an object
  const selections: FieldNode[] = [];

  Object.keys(obj).forEach(key => {
    const field: FieldNode = {
      kind: 'Field',
      name: {
        kind: 'Name',
        value: key,
      },
    };

    // Recurse
    const nestedSelSet: SelectionSetNode = selectionSetFromObj(obj[key]);

    if (nestedSelSet) {
      field.selectionSet = nestedSelSet;
    }

    selections.push(field);
  });

  const selectionSet: SelectionSetNode = {
    kind: 'SelectionSet',
    selections,
  };

  return selectionSet;
}

export const justTypenameQuery: DocumentNode = {
  kind: 'Document',
  definitions: [
    {
      kind: 'OperationDefinition',
      operation: 'query',
      name: null,
      variableDefinitions: null,
      directives: [],
      selectionSet: {
        kind: 'SelectionSet',
        selections: [
          {
            kind: 'Field',
            alias: null,
            name: {
              kind: 'Name',
              value: '__typename',
            },
            arguments: [],
            directives: [],
            selectionSet: null,
          },
        ],
      },
    },
  ],
};
