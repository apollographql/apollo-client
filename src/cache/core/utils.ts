import {
  DocumentNode,
  SelectionSetNode,
} from 'graphql';

export function queryFromPojo(obj: any): DocumentNode {
  return {
    kind: 'Document',
    definitions: [{
      kind: 'OperationDefinition',
      operation: 'query',
      name: {
        kind: 'Name',
        value: 'GeneratedClientQuery',
      },
      selectionSet: selectionSetFromObj(obj),
    }],
  };
}

export function fragmentFromPojo(obj: any, typename?: string): DocumentNode {
  return {
    kind: 'Document',
    definitions: [{
      kind: 'FragmentDefinition',
      typeCondition: {
        kind: 'NamedType',
        name: {
          kind: 'Name',
          value: typename || '__FakeType',
        },
      },
      name: {
        kind: 'Name',
        value: 'GeneratedClientQuery',
      },
      selectionSet: selectionSetFromObj(obj),
    }],
  };
}

function selectionSetFromObj(obj: any): SelectionSetNode {
  if (!obj || Object(obj) !== obj) {
    // No selection set here
    return null;
  }

  if (Array.isArray(obj)) {
    // GraphQL queries don't include arrays
    return selectionSetFromObj(obj[0]);
  }

  // Now we know it's an object
  return {
    kind: 'SelectionSet',
    selections: Object.keys(obj).map(key => ({
      kind: 'Field',
      name: {
        kind: 'Name',
        value: key,
      },
      selectionSet: selectionSetFromObj(obj[key]) || void 0,
    })),
  };
}
