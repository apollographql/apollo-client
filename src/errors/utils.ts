export function isBranded(error: unknown, name: string) {
  return (
    typeof error === "object" &&
    error !== null &&
    (error as any)[Symbol.for("apollo.error")] === name
  );
}

export function brand(error: Error) {
  Object.defineProperty(error, Symbol.for("apollo.error"), {
    value: error.name,
    enumerable: false,
    writable: false,
    configurable: false,
  });
}
