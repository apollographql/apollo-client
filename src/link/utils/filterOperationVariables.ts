import type { VariableDefinitionNode, DocumentNode } from "graphql";
import { visit } from "graphql";

export function filterOperationVariables(
  variables: Record<string, any>,
  query: DocumentNode
) {
  const result = { ...variables };
  const unusedNames = new Set(Object.keys(variables));
  visit(query, {
    Variable(node, _key, parent) {
      // A variable type definition at the top level of a query is not
      // enough to silence server-side errors about the variable being
      // unused, so variable definitions do not count as usage.
      // https://spec.graphql.org/draft/#sec-All-Variables-Used
      if (
        parent &&
        (parent as VariableDefinitionNode).kind !== "VariableDefinition"
      ) {
        unusedNames.delete(node.name.value);
      }
    },
  });
  unusedNames.forEach((name) => {
    delete result![name];
  });
  return result;
}
