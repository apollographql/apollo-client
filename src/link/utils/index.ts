export { fromError } from './fromError';
export { toPromise } from './toPromise';
export { fromPromise } from './fromPromise';
export { ServerError, throwServerError } from './throwServerError';
export { validateOperation } from './validateOperation';
export { createOperation } from './createOperation';
export { transformOperation } from './transformOperation';
