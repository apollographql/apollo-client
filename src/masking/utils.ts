import { Slot } from "optimism";
import { invariant } from "../utilities/globals/index.js";
import { canUseWeakMap, canUseWeakSet } from "../utilities/index.js";

export const MapImpl = canUseWeakMap ? WeakMap : Map;
export const SetImpl = canUseWeakSet ? WeakSet : Set;

// Contextual slot that allows us to disable accessor warnings on fields when in
// migrate mode.
/** @internal */
export const disableWarningsSlot = new Slot<boolean>();

let issuedWarning = false;
export function warnOnImproperCacheImplementation() {
  if (!issuedWarning) {
    issuedWarning = true;
    invariant.warn(
      "The configured cache does not support data masking which effectively disables it. Please use a cache that supports data masking or disable data masking to silence this warning."
    );
  }
}
