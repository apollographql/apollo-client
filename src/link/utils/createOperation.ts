import { GraphQLRequest, Operation } from '../core';

export function createOperation(
  starting: any,
  operation: GraphQLRequest,
): Operation {
  let context = { ...starting };
  const setContext = (next: any) => {
    if (typeof next === 'function') {
      context = { ...context, ...next(context) };
    } else {
      context = { ...context, ...next };
    }
  };
  const getContext = () => ({ ...context });

  Object.defineProperty(operation, 'setContext', {
    enumerable: false,
    value: setContext,
  });

  Object.defineProperty(operation, 'getContext', {
    enumerable: false,
    value: getContext,
  });

  return operation as Operation;
}
