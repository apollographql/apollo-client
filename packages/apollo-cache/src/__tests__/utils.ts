import { print } from 'graphql/language/printer';
import { queryFromPojo, fragmentFromPojo } from '../utils';

describe('writing data with no query', () => {
  describe('converts a JavaScript object to a query correctly', () => {
    it('basic', () => {
      expect(
        print(
          queryFromPojo({
            number: 5,
            bool: true,
            bool2: false,
            undef: undefined,
            nullField: null,
            str: 'string',
          }),
        ),
      ).toMatchSnapshot();
    });

    it('nested', () => {
      expect(
        print(
          queryFromPojo({
            number: 5,
            bool: true,
            nested: {
              bool2: false,
              undef: undefined,
              nullField: null,
              str: 'string',
            },
          }),
        ),
      ).toMatchSnapshot();
    });

    it('arrays', () => {
      expect(
        print(
          queryFromPojo({
            number: [5],
            bool: [[true]],
            nested: [
              {
                bool2: false,
                undef: undefined,
                nullField: null,
                str: 'string',
              },
            ],
          }),
        ),
      ).toMatchSnapshot();
    });

    it('fragments', () => {
      expect(
        print(
          fragmentFromPojo({
            number: [5],
            bool: [[true]],
            nested: [
              {
                bool2: false,
                undef: undefined,
                nullField: null,
                str: 'string',
              },
            ],
          }),
        ),
      ).toMatchSnapshot();
    });
  });
});
