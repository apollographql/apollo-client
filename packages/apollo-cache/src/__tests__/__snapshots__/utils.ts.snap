// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[
  `writing data with no query converts a JavaScript object to a query correctly arrays 1`
] = `
"query GeneratedClientQuery {
  number
  bool
  nested {
    bool2
    undef
    nullField
    str
  }
}
"
`;

exports[
  `writing data with no query converts a JavaScript object to a query correctly basic 1`
] = `
"query GeneratedClientQuery {
  number
  bool
  bool2
  undef
  nullField
  str
}
"
`;

exports[
  `writing data with no query converts a JavaScript object to a query correctly fragments 1`
] = `
"fragment GeneratedClientQuery on __FakeType {
  number
  bool
  nested {
    bool2
    undef
    nullField
    str
  }
}
"
`;

exports[
  `writing data with no query converts a JavaScript object to a query correctly nested 1`
] = `
"query GeneratedClientQuery {
  number
  bool
  nested {
    bool2
    undef
    nullField
    str
  }
}
"
`;
